### PR TITLE
Finishing touches on @bhack's SliceLayer (#680)

### PR DESCRIPTION
**Pull request by jeffdonahue**
Finishes #680. Thanks for contributing this layer @bhack, and thanks for your review @sguada.  Will merge once Travis passes.

### CODE REVIEW

File: include/caffe/common_layers.hpp [hunk 1]
```diff
@@ -178,6 +178,44 @@ class SplitLayer : public Layer<Dtype> {
...
+  explicit SliceLayer(const LayerParameter& param)
+      : Layer<Dtype>(param) {}
+  virtual void SetUp(const vector<Blob<Dtype>*>& bottom,
+      vector<Blob<Dtype>*>* top);
+
+  virtual inline LayerParameter_LayerType type() const {
+    return LayerParameter_LayerType_SLICE;
+  }
+  virtual inline int ExactNumBottomBlobs() const { return 1; }
+  virtual inline int MinTopBlobs() const { return 2; }
```
Comment (ronghanghu) [inline include/caffe/common_layers.hpp]: I am sort of confused here when I use this layer, and traced back to here... Why `MinTopBlobs()` returns 2? (why not 1?)

Comment (bhack) [inline include/caffe/common_layers.hpp]: @ronghanghu Slice is a knife for a blob. A division by 1 exist in math. But is it operative useful when you are cooking?

Comment (ronghanghu) [inline include/caffe/common_layers.hpp]: You are right :)

I was trying to use slice layer as a component in another layer to automatically slice T x ... blob into T blobs of 1x ..., and run into this error when T == 1. I'll handle the special case myself.

Comment (seanbell) [inline include/caffe/common_layers.hpp]: Why not allow for 1 top blob?  I understand that it doesn't make sense to use a `SliceLayer` for that purpose, but if a user asks for it, the layer could act as a pass-through, doing nothing at all.  This makes the layer more general.

Comment (bhack) [inline include/caffe/common_layers.hpp]: If there are some derived reuse of this layer then could be extended to hande the Top 1 case in a way that not impact performance (cause it is a useless computation) so the logic doesn't need to be reproduced everywhere.

**Pull request merged**